nautobot_device_lifecycle_mgmt/jobs/cve_tracking.py

```diff
@@ -58,3 +71,391 @@ def run(self, published_after, debug=False):  # pylint: disable=too-many-locals,
+        """Search NIST for software and related CVEs.
+
+        Args:
+            cpe_software_search_urls (list): List of URLs to query for CVE information.
+            software (object): Software object being queried.
+
+        Returns:
+            dict: Dictionary containing new and existing CVE information.
+        """
+        cve_list = []
+        all_cve_info = {"new": {}, "existing": {}}
+
+        for cpe_software_search_url in cpe_software_search_urls:
+            result = self.query_api(cpe_software_search_url)
+            if result["totalResults"] > 0:
+                cve_list.extend([cve for cve in result["vulnerabilities"] if cve not in cve_list])
+        if cve_list:
+            self.logger.info(
+                "Received %s results.",
+                len(cve_list),
+                extra={"object": software, "grouping": "CVE Creation"},
+            )
+            all_cve_info = self.process_cves(cve_list, software)
+
+        return all_cve_info
+
+    def process_cves(self, cve_list: list, software: SoftwareVersion) -> dict:
+        """Return processed CVE info categorized as new or existing.
+
+        Args:
+            cve_list (list): List of CVEs returned from CPE search.
+            software (object): Software object being queried.
+
+        Returns:
+            dict: Dictionary of CVEs categorized as new or existing.
+        """
+        processed_cve_info = {"new": {}, "existing": {}}
+        dlc_cves = CVELCM.objects.values_list("name", flat=True)
+
+        for cve in cve_list:
+            if not cve["cve"]["id"].startswith("CVE"):
+                continue
+            if cve["cve"]["id"] not in dlc_cves:
+                processed_cve_info["new"].update({cve["cve"]["id"]: self.prep_cve_for_dlc(cve["cve"])})
+            else:
+                processed_cve_info["existing"].update({cve["cve"]["id"]: self.prep_cve_for_dlc(cve["cve"])})
+        self.logger.info(
+            "Prepared %s CVE for creation." % len(processed_cve_info["new"]),
+            extra={"object": software, "grouping": "CVE Creation"},
+        )
+
+        return processed_cve_info
+
+    def query_api(self, url: str) -> dict:
+        """Query the NIST API using an established session.
+
+        Args:
+            url (str): The API endpoint being queried.
+
+        Returns:
+            dict: Dictionary of returned results if successful.
+        """
+        # Build retry configuration; Uses Integration settings but setting sane defaults in case they are removed
+        retries = Retry(
+            total=self.integration.extra_config.get("retries", {}).get("max_attempts", 3),
+            backoff_factor=self.integration.extra_config.get("retries", {}).get("backoff", 1),
+            status_forcelist=[502, 503, 504],
+            allowed_methods=["GET"],
+        )
+
+        # Create and configure session
+        session = Session()
```

Comment:
From the docstring "Query the NIST API using an established session" -> why are we creating a new `Session` every time?

Reply:
I believe the docstring is poorly worded.  The NVD doesn't allow retrieving multiple CVE in a single call.

This is meant to create the session so that retries can be defined/used per CVE retrieval.  Let me try to update the docstring to be a bit more clear on what is happening or just remove the "using an established session" part.

Reply:
I just took the approach of removing the "using an established session" part.

Reply:
Actually, we should init the session once in the `run()`. You can make it an instance attribute, such as `self.nist_session=self.nist_session_init()`.

Reply:
I moved it to the run as requested.

I remembered that at one time there was an issue when sending the call for CVE info was failing at random and out of X amount of calls there might be a lot of them that would fail and need a retry, or multiple retries, before that CVE would be retrieved.  That's why there was a session being created for each one.

I hope moving it to a single session with a specific number of retries does not cause issues in those cases but I have been unable to reproduce for a while now.

Maybe it was just an issue with the NVD that has been resolved.